src/mesh/analysis.rs

```diff
@@ -1,13 +1,125 @@
-use std::collections::HashSet;
+use std::collections::{HashMap, HashSet};
 
-use nalgebra::base::Vector3;
-use nalgebra::geometry::Point3;
+use nalgebra as na;
+use nalgebra::{Point3, Vector3};
 
 use crate::convert::{cast_i32, cast_usize};
-use crate::edge_analysis::EdgeSharingMap;
-use crate::geometry::{Face, Geometry, OrientedEdge, UnorientedEdge};
 
-/// Finds edges with a certain valency in a mesh edge collection
+use super::{Face, Mesh, OrientedEdge, UnorientedEdge};
+
+pub fn compute_bounding_sphere<'a, I>(meshes: I) -> (Point3<f32>, f32)
+where
+    I: IntoIterator<Item = &'a Mesh> + Clone,
+{
+    let centroid = compute_centroid(meshes.clone());
+    let mut max_distance_squared = 0.0;
+
+    for mesh in meshes {
+        for vertex in &mesh.vertices {
+            let distance_squared = na::distance_squared(&centroid, vertex);
+            if distance_squared > max_distance_squared {
+                max_distance_squared = distance_squared;
+            }
+        }
+    }
+
+    (centroid, max_distance_squared.sqrt())
+}
+
+pub fn compute_centroid<'a, I>(meshes: I) -> Point3<f32>
+where
+    I: IntoIterator<Item = &'a Mesh>,
+{
+    let mut vertex_count = 0;
+    let mut centroid = Point3::origin();
+    for mesh in meshes {
+        vertex_count += mesh.vertices.len();
+        for vertex in &mesh.vertices {
+            let v = vertex - Point3::origin();
+            centroid += v;
+        }
+    }
+
+    centroid / (vertex_count as f32)
+}
+
+// FIXME: deprecate and remove, superseded by pull point to mesh
+pub fn find_closest_point(position: &Point3<f32>, mesh: &Mesh) -> Option<Point3<f32>> {
+    let vertices = mesh.vertices();
+    if vertices.is_empty() {
+        return None;
+    }
+
+    let mut closest = vertices[0];
+    let mut closest_distance_squared = na::distance_squared(position, &closest);
+    for point in &vertices[1..] {
+        let distance_squared = na::distance_squared(position, &point);
+        if distance_squared < closest_distance_squared {
+            closest = *point;
+            closest_distance_squared = distance_squared;
+        }
+    }
+
+    Some(closest)
+}
+
+/// The edges sharing the same vertex indices.
+/// ascending_edges contains edges oriented from lower index to higher
+/// descending_edges contains edges oriented from higher index to lower
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SharedEdges {
```

Comment:
Not sure, perhaps you had a reason for keeping it here, but it seems to me this semantically belongs to `topology.rs`. It applies to all the other edge sharing functions.

Reply:
Hmmm, yes, it is a bit similar. I can try moving it there to see if any dependencies break.

Reply:
Attempted this, but the result was not very convincing - there's a lot of dependencies on the edge sharing in analysis. I'd leave it be for now and we can revisit later.